### DISCUSSION TITLE
h264_v4l2m2m rpi

### DISCUSSION THREAD

**Discussion by spookyfirehorse** [bluenviron/mediamtx#1282] [General]
Hi and Hallo !

sorry for my bad english !!!

computer rpi 4 and rpi 3

boot config

start_x=1
#camera_auto_detect=1

ffmpeg 4.3.5

and thats my script

ffmpeg -hwaccel vulkan -fflags +genpts+nobuffer+igndts+discardcorrupt -hide_banner -strict experimental -async 1
-f alsa -ac 2 -i hw:CARD=S3,DEV=0 -f v4l2 -input_format h264 -use_wallclock_as_timestamps 1 -re -i /dev/video0 -c:v h264_v4l2m2m -b:v 1M -pix_fmt yuv420p
-c:a libopus -b:a 32k -application lowdelay -movflags +faststart+write_colr
-f rtsp -rtsp_transport tcp rtsp://localhost:8557/mystream

the stream is running but when i was looking for that strem with

mpv rtsp://spook:password@lucy:8557/mystream

error massege

[ffmpeg/video] h264: non-existing PPS 0 referenced
[ffmpeg/video] h264: non-existing PPS 0 referenced
[ffmpeg/video] h264: decode_slice_header error
[ffmpeg/video] h264: no frame!

stream is running with ffmpeg 4.3.2

and this day libcamera was updated by raspbian libcamera.so.0.1 to libcamera.so.0.2
so rtsp-simple-server searching for libcamera.so.0.1 error

thank you for your great work

**Comment by aler9** [comment]
Hello, it turned out that FFmpeg 4.3.2 had a bug that prevented SPS and PPS from being sent to the server when using h264_v4l2m2m (and that's the reason of the "non-existing PPS 0 referenced" message). Solution is here: https://github.com/aler9/rtsp-simple-server/issues/1155#issuecomment-1371494056

**Comment by spookyfirehorse** [comment]
Hi thank you for your great work and the answer but with ffmeg 5.1.2 the
same error.


Alessandro Ros ***@***.***> schrieb am Mi., 4. Jän. 2023,
23:19:

**Reply by aler9** [reply]
@spookyfirehorse i used the following command in order to develop the fix:

```
ffmpeg -re -f lavfi -i testsrc=duration=300:size=1280x720:rate=30 -c:v h264_v4l2m2m -b:v 1M -f rtsp rtsp://localhost:8554/stream
```

this was not generating SPS/PPS with FFmpeg 4.3.5, while it's generating SPS/PPS with FFmpeg 5.1.1.
Anyway, the server is only responsible for receiving data from FFmpeg, it doesn't influence how the data is generated. If problem persists, it's a FFmpeg issue.

**Comment by spookyfirehorse** [comment]
Thank you i ASK jc-kynesim whats going on

Alessandro Ros ***@***.***> schrieb am Do., 5. Jän. 2023,
18:03:

**Comment by spookyfirehorse** [comment]
i used this comand

just copy

this works

v4l2-ctl -d /dev/video0  -p 25 
--set-fmt-video=width=640,height=360,pixelformat=4 
--set-ctrl=brightness=57,contrast=-11,exposure_dynamic_framerate=0,h264_level=11,h264_profile=2,video_bitrate=10000000,h264_i_frame_period=25

ffmpeg  -fflags +genpts+nobuffer+igndts+discardcorrupt -hide_banner  
-strict experimental -async 1  \
   -f alsa  -ac 1  -i hw:CARD=Device,DEV=0  -f v4l2 -input_format h264 
-use_wallclock_as_timestamps 1  -re  -f v4l2  -i /dev/video0 -c:v copy  
-pix_fmt yuv420p    \
   -c:a libopus  -b:a 32k  -application lowdelay   \
   -f rtsp -rtsp_transport tcp rtsp://localhost:8554/mystream


thank you for your comand but the problem not go away

[a comment by an automated account is not shown]
